languages/python/exercises/concept/basics/.docs/introduction.md

```diff
@@ -0,0 +1,95 @@
+Python is a [dynamic and strongly](https://stackoverflow.com/questions/11328920/is-python-strongly-typed) typed [object-oriented](https://en.wikipedia.org/wiki/Object-oriented_programming) programming language. It supports both imperative (_object-oriented, procedural_) and declarative (_functional, concurrent_) programming paradigms. Python emphasizes code readability and (similar to Haskell) uses significant indentation.
+
+Objects are [assigned](https://docs.python.org/3/reference/simple_stmts.html#assignment-statements) to names in Python via the `=` or _assignment operator_. _Variables_ are written in [`snake_case`](https://en.wikipedia.org/wiki/Snake_case), and _constants_ (usually) in `SCREAMING_SNAKE_CASE`. A name (_variable or constant_) is not itself _typed_, and can be attached or re-attached to different objects over its lifetime:
+
+```python
+>>> my_first_variable = 1
+>>> my_first_variable = {"bear" : "brown", "cat": "black"}
+>>> my_first_variable = collections.Counter()
+>>> my_first_variable = "Last one, I promise"
+>>> print(my_first_variable)
+"Last one, I promise"
+```
+
+Constants are usually defined at the top of a script or file, and although they _can_ be changed, they are _intended_ to be named only once. Their `SCREAMING_SNAKE_CASE` is a message to other developers that the assignment should not be altered:
+
+```python
+# All caps signal that this is intended as a constant
+MY_FIRST_CONSTANT = 16
+
+# Re-assignment will be allowed by the compiler & interpreter,
+# but is VERY strongly discouraged.
+# Please don't do: MY_FIRST_CONSTANT = "Some other value"
+```
+
+In Python, units of functionality are encapsulated in [functions](https://docs.python.org/3/reference/compound_stmts.html#function). When functions are bound to a [class](https://docs.python.org/3/reference/datamodel.html#classes) name, they're referred to as [methods](https://docs.python.org/3/c-api/method.html#method-objects).
+
+The keyword `def` begins a [function definition](https://docs.python.org/3/tutorial/controlflow.html#defining-functions). It must be followed by the function name and a parenthesized list of zero or more formal [parameters](https://docs.python.org/3/glossary.html#term-parameter). The `def` line is terminated with a colon. Statements for the _body_ of the function begin on the next line, and must be _indented in a block_. Functions explicitly return a value or object via the `return` keyword:
```

Comment:
Python is white-space sensitive, right? I don't know any Python, but I was wondering if it would be helpful to explain how to indent properly (can I use tabs/spaces?, how many tabs/spaces?) as I feel that is something that might trip up new students.

Reply:
> Python is white-space sensitive, right? 

Correct.

> can I use tabs/spaces?, how many tabs/spaces?

I couldn't find a more recent version of this doc, but it's still correct: https://docs.python.org/2.0/ref/indentation.html

TLDR: You can use either, but be consistent and readable. Mixing tabs and spaces in the same file is a `SyntaxError`

@ErikSchierboom On a related note: has any other track done an exercise on linters and formatters? It's quite common to at least use `flake8` or `pylint` to check your Python code for style violations and `SyntaxError`s, and there are a few good formatters out there as well (`black` and `autopep8` to name a couple).

Reply:
@cmccandless We haven't, because linters and formatters are usually not part of the language. I think having an exercise for them makes less sense, but it would definitely be something that we can include in instructions for CLI users (we have a [special file for this in v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#exercisesshareddocsclimd), which will automatically apply to all exercises in a track).

Reply:
**PENDANT ALERT**  😉 

> Python is white-space sensitive, right?

The reason I used "significant indent" is because _some_ whitespace matters in Python ..but other whitespace doesn't -- or is implicitly concatenated by the interpreter.  The [lexical analysis docs](https://docs.python.org/3/reference/lexical_analysis.html#line-structure) draw distinctions between _logical lines_ and _physical lines_.  It can be a bit annoying.  For example, whitespace in a dictionary literal, tuple, list, or comprehension is implicitly concatenated.  But if the declaration appears inside a function or class and you _don't_ indent the logical line, then you get an `IndentationError`: 

```python
#this is implicitly concatenated by the interpreter
indented_dict = {
                 "cat" : "orange tabby", 
                 "dog": "brown beagle",
                 "bird": "blue mockingjay"
                }

#this is too
indented_dict_two = {"cat" : "orange tabby", 
                     "dog": "brown beagle",
                     "bird": "blue mockingjay"}

#this is perfectly fine, and won't toss any errors.
one_line_dict = {"cat" : "orange tabby", "dog": "brown beagle", "bird": "blue mockingjay"}

#but this will toss an error, because the dict variable name isn't indented properly.
def example_function():
    example_variable = "this is a string."
  indented_dict = {
                 "cat" : "orange tabby", 
                 "dog": "brown beagle",
                 "bird": "blue mockingjay"
                }
```

Pedantry aside, I added this entry on [indentation](https://docs.python.org/3/reference/lexical_analysis.html#indentation) to the `after.md` (_looks like Python 3.8 now has a `TabError`...a subclass of `IndentationError`.  lol_), and added the following to the `introduction.md`:

>  Statements for the _body_ of the function begin on the next line, and must be _indented in a block_. There is no strict indentation amount (_either space **OR** [tab] characters are acceptable_), but indentation must be _consistent for all indented lines_.